### PR TITLE
Cast from numeric/timestamp to timestamp/numeric

### LINKED ISSUE #5122: Cast from floating/timestamp to timestamp/floating
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Similar to #5039, floating types cannot be casted to/from timestamp types for now. This is supported by Spark. We'd like to fill this gap too.

**Describe the solution you'd like**

Supporting casting from floating/timestamp to timestamp/floating by cast kernel.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #5122.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -4719,6 +4726,35 @@ mod tests {
         assert_eq!(&actual, &expected);
     }
 
+    #[test]
+    fn test_cast_floating_to_timestamp() {
+        let array = Int64Array::from(vec![Some(2), Some(10), None]);
+        let expected = cast(&array, &DataType::Timestamp(TimeUnit::Microsecond, None)).unwrap();
+
+        let array = Float32Array::from(vec![Some(2.0), Some(10.6), None]);
```
Comment (viirya) [inline arrow-cast/src/cast.rs:4734]: I don't add test for Float16Array as casting between Float16 and Int64 is not supported by cast kernel now. I will make it in a separate PR later.

File: arrow-cast/src/cast.rs [hunk 2]
```diff
@@ -1634,24 +1634,31 @@ pub fn cast_with_options(
                 .unary::<_, Time64MicrosecondType>(|x| x / (NANOSECONDS / MICROSECONDS)),
         )),
 
-        (Timestamp(TimeUnit::Second, _), _) if to_type.is_integer() => {
+        // Timestamp to integer/floating
+        (Timestamp(TimeUnit::Second, _), _) if to_type.is_integer() || to_type.is_floating() => {
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:1638]: Perhaps we could use `is_numeric` or is there a particular reason to exclude decimals?

**Review by tustvold — APPROVED** [review]

File: arrow-cast/src/cast.rs (hunk 2 above)
Comment (viirya) [reply arrow-cast/src/cast.rs:1638]: No, I plan to add decimals too (in a separate PR).

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
Seems to be a slight issue with can_cast_types

**Pull request merged**